moz_kinto_publisher/main.py

```diff
@@ -961,6 +961,136 @@ def publish_crlite(*, args, ro_client, rw_client):
             )
 
 
+def publish_ctlogs(*, args, ro_client, rw_client):
+    # Copy CT log metadata from google's v3 log_list to our kinto collection.
+    # This will notify reviewers who can then manually enroll the log in CRLite.
+    #
+    # Schema for our ct-logs kinto collection:
+    #   {
+    #       "crlite_enrolled": boolean,
+    #       "description": string,
+    #       "key": string,
+    #       "log_id": string,
+    #       "mmd": integer,
+    #       "url": string
+    #   }
+    #
+
+    log_list_json = requests.get(
+        "https://www.gstatic.com/ct/log_list/v3/log_list.json"
+    ).json()
+
+    # Google groups CT logs according to their operators, we want a flat list
+    upstream_logs_raw = [
+        ctlog for operator in log_list_json["operators"] for ctlog in operator["logs"]
+    ]
+
+    # Translate |upstream_logs_raw| to our schema (and remove unused fields)
+    upstream_logs = [
+        {
+            "crlite_enrolled": False,
+            "description": ctlog["description"],
+            "key": ctlog["key"],
+            "logID": ctlog["log_id"],
+            "url": ctlog["url"],
+            "mmd": ctlog["mmd"],
+        }
+        for ctlog in upstream_logs_raw
+    ]
+    upstream_lut = {ctlog["logID"]: ctlog for ctlog in upstream_logs}
+
+    if len(upstream_logs) != len(upstream_lut):
+        raise SanityException(
+            "We expect the 'log_id' field to be unique in log_list.json"
+        )
+
+    # LogID is supposed to be a hash of the CT Log's key
+    for upstream_log in upstream_logs:
+        rfc6962_log_id = base64.b64encode(
+            hashlib.sha256(base64.b64decode(upstream_log["key"])).digest()
+        )
+        if rfc6962_log_id != upstream_log["logID"].encode("utf8"):
+            raise SanityException(
+                f"Google log list contains incorrectly computed logID {upstream_log}"
+            )
+
+    # Fetch our existing Kinto records
+    known_logs = ro_client.get_records(collection=settings.KINTO_CTLOGS_COLLECTION)
+    known_lut = {ctlog["logID"]: ctlog for ctlog in known_logs}
```

Comment:
What does `lut` stand for?
Also, do we want to consistency-check the preexisting kinto data, just in case we've put something bad in there?

Reply:
"Lookup table". Open to suggestions.

I added a test to ensure that the logID fields are unique. With this change, 1) each unique logID in our list corresponds to a unique entry in Google's list; and 2) the "description", "key", "url", and "mmd" fields (i.e. everything but "crlite_enrolled") match the corresponding entries in Google's list. The only field not covered is "crlite_enrolled", which doesn't need any consistency checks.

Reply:
Ok - looks good to me.